File: ex1019.java
// Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}


class Solution {
    public int[] nextLargerNodes(ListNode head) {
        int cnt = 0;
        for (ListNode p = head; p != null; p = p.next)    cnt++;
        
        int[] arr = new int[cnt];
        int a = 0;
        for (ListNode p = head; p != null; p = p.next)    arr[a++] = p.val;
        
        int[] res = new int[cnt];
        for (int i = 0; i < cnt; i++) {
            for (int j = i + 1; j < cnt; j++) {
                if (arr[j] > arr[i]) {
                    res[i] = arr[j];
                    j = cnt;
                }
            }
        }
        
        return res;
    }
}

File: code/ex1019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        stack<pair<int, int>> stk;    // {val, idx}
        vector<int> res;
        
        for (ListNode* p = head; p != NULL; p = p->next) {
            res.push_back(0);
            while (!stk.empty() && stk.top().first < p->val) {
                res[stk.top().second] = p->val;
                stk.pop();
            }
            
            stk.push(pair(p->val, res.size() - 1));
        }
        
        return res;
    }
};